src/Extension/TemplateExtension.php

```diff
@@ -31,38 +33,129 @@ final class TemplateExtension extends AbstractExtension
     protected $modelAdapter;
 
     /**
-     * @param bool             $debug        Is Symfony debug enabled?
-     * @param AdapterInterface $modelAdapter A Sonata model adapter
+     * NEXT_MAJOR: remove this property.
+     *
+     * @var LegacyTranslatorInterface|TranslatorInterface|null
+     *
+     * @deprecated translator property is deprecated since sonata-project/twig-extensions 0.x, to be removed in 1.0
      */
-    public function __construct(bool $debug, AdapterInterface $modelAdapter)
+    protected $translator;
+
+    /**
+     * @param bool                                                                $debug                              Is Symfony debug enabled?
+     * @param LegacyTranslatorInterface|TranslatorInterface|AdapterInterface|null $deprecatedTranslatorOrModelAdapter
+     */
+    public function __construct($debug, $deprecatedTranslatorOrModelAdapter = null, ?AdapterInterface $deprecatedModelAdapter = null)
     {
         $this->debug = $debug;
-        $this->modelAdapter = $modelAdapter;
+
+        if (
+            !$deprecatedTranslatorOrModelAdapter instanceof LegacyTranslatorInterface &&
+            !$deprecatedTranslatorOrModelAdapter instanceof TranslatorInterface &&
+            !$deprecatedTranslatorOrModelAdapter instanceof AdapterInterface &&
+            null !== $deprecatedTranslatorOrModelAdapter
+        ) {
+            throw new \InvalidArgumentException(sprintf(
```

Comment:
If we keep the ability to accept multiple types in the next major, this exception should become of type `\TypeError`.